_posts/2021/2021-09-28-cpp-lessons-learned.md

```diff
@@ -0,0 +1,147 @@
+* **Figure 2.** Radix_tree with inline_string - only one allocation.
+
+It's one but bigger allocation, so it suffers when it comes to updating the string data.
+In case when the value is longer than the existing string allocation - **the whole node**
+has to be re-allocated.
+
+
+#### Using per-thread data
+
+Very recently, we've described ([in the "Concurrency considerations" blog post][concurrency_blog])
+our approach to ensure data consistency. It's a very challenging task, especially if
+you consider concurrent modifications within a container full of records, with multiple
+threads accessing/modifying the same structure. Read the whole story for details,
+but long story short - per-thread data (implemented in libpmemobj-cpp as
+[enumerable_thread_specific][ptls_docs]) can save you from data consistency issues.
+
+
+#### Keeping locks on pmem is not a good idea
+
+Since we keep all data for our application on persistent memory, one might think we should
+do that as well with locks. That's the wrong way. Each lock and unlock operation might end up
+with a write to pmem. Our understanding is that, even though we do not explicitly flush cache
+lines, they are invalidated when a lock is accessed from a different core. Additional write
+to pmem is, of course, undesired and limits the scalability of our data structures. There are
+two solutions we propose to fix this problem:
+
+The first one is to keep per-element locks in DRAM instead (pointers to these locks are held
+on pmem). Mentioned additional writes are less costly to DRAM than to persistent memory.
+Currently, within our library, it is only possible with `concurrent_hash_map`.
+It's a good approach if you store large elements.
+
+The second solution is to use sharding instead of per-element locks. Data can be divided
+into multiple single-threaded buckets, each protected by a lock. With a uniform distribution
+of elements and a big enough number of buckets, this solution will provide a short waiting
+time on locks and good performance. In practice, this can be easily implemented, e.g., in
+unsorted data structures. It's possible to mix this solution with the previous one for
+even further optimization.
+
+
+#### Lazy rehashing
+
+Usually, in a structure with hashing table, it is required to perform, from time to time,
+a full rehashing of existing entries. In a concurrent environment, this task is even more
+challenging and may require locking the entire structure. With this approach, you will
+observe latency spikes in moments of such rehashing, slowing down each new write to the
+structure. In our `concurrent_hash_map` we introduced a "lazy rehashing" solution, based
+on the paper ["Per-bucket concurrent rehashing algorithms"][rehashing_pdf]. When rehashing
+is required, the number of buckets is doubled. Later, while accessing (in any way)
+an element, it may be moved into some of the new buckets or may stay in place. No locks
```

Comment:
Actually there are per-bucket locks when we move elements, but there are no global locks taken.